mm2src/coins/eth/web3_transport/websocket_transport.rs

```diff
@@ -0,0 +1,424 @@
+    }
+
+    async fn handle_send_request(
+        &self,
+        request: Option<ControllerMessage>,
+        wsocket: &mut WebSocketStream,
+        response_notifiers: &mut ExpirableMap<usize, oneshot::Sender<()>>,
+    ) -> OuterAction {
+        match request {
+            Some(ControllerMessage::Request(WsRequest {
+                request_id,
+                serialized_request,
+                response_notifier,
+            })) => {
+                response_notifiers.insert(
+                    request_id,
+                    response_notifier,
+                    Duration::from_secs(REQUEST_TIMEOUT_AS_SEC),
+                );
+
+                let mut should_continue = Default::default();
+                for _ in 0..MAX_ATTEMPTS {
+                    match wsocket
+                        .send(tokio_tungstenite_wasm::Message::Text(serialized_request.clone()))
+                        .await
+                    {
+                        Ok(_) => {
+                            should_continue = false;
+                            break;
+                        },
+                        Err(e) => {
+                            log::error!("{e}");
+                            should_continue = true;
+                        },
+                    }
+
+                    Timer::sleep(SLEEP_DURATION).await;
+                }
+
+                if should_continue {
+                    let _ = response_notifiers.remove(&request_id);
+                    return OuterAction::Continue;
+                }
+            },
+            Some(ControllerMessage::Close) => {
+                return OuterAction::Break;
+            },
+            _ => {},
+        }
+
+        OuterAction::None
+    }
+
+    async fn handle_response(
+        &self,
+        message: Option<Result<tokio_tungstenite_wasm::Message, tokio_tungstenite_wasm::Error>>,
+        response_notifiers: &mut ExpirableMap<usize, oneshot::Sender<()>>,
+    ) -> OuterAction {
+        match message {
+            Some(Ok(tokio_tungstenite_wasm::Message::Text(inc_event))) => {
+                if let Ok(inc_event) = serde_json::from_str::<serde_json::Value>(&inc_event) {
+                    if !inc_event.is_object() {
+                        return OuterAction::Continue;
+                    }
+
+                    if let Some(id) = inc_event.get("id") {
+                        let request_id = id.as_u64().unwrap_or_default() as usize;
+
+                        if let Some(notifier) = response_notifiers.remove(&request_id) {
+                            let mut res_bytes: Vec<u8> = Vec::new();
+                            if serde_json::to_writer(&mut res_bytes, &inc_event).is_ok() {
+                                let response_map = unsafe { &mut *self.responses.0 };
+                                let _ = response_map.insert(request_id, res_bytes);
+
+                                notifier.send(()).expect("receiver channel must be alive");
```

Comment:
Won't this ever panic?

Reply:
No. Caller is awaiting for the response through this channel. If it ever drop before that, we should panic as something def. going wrong.

Reply:
ahaa I was thinking this was a notifier for external communication (not us).